src/mame/msx/msx.cpp

```diff
@@ -453,6 +455,35 @@ void msx_state::kanji_w(offs_t offset, u8 data)
 		m_kanji_latch = (m_kanji_latch & 0x1f800) | ((data & 0x3f) << 5);
 }
 
+u8 msx_state::kanji2_r(offs_t offset)
+{
+	u8 result = 0xff;
+
+	if (m_region_kanji && m_region_kanji->bytes() > 0x20000)
+	{
+		// TODO: Are there one or two latches in a system?
+		u32 latch = m_kanji_fsa1fx ? bitswap<17>(m_kanji_latch, 4, 3, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 2, 1, 0) : m_kanji_latch;
+		result = m_region_kanji->as_u8(0x20000 | latch);
+
+		if (!machine().side_effects_disabled())
+		{
+			m_kanji_latch = (m_kanji_latch & ~0x1f) | ((m_kanji_latch + 1) & 0x1f);
+		}
+	}
+	return result;
+}
```

Comment:
Can you turn `m_kanji_region` into an `optional_region_ptr<u8>` and index it directly (plus get some more validation) rather than needing `as_u8`?

Should you take the `address_space &` parameter and return `space.unmap()` if the kanji ROM is not present or too small, or is it guaranteed to read high?

Or alternatively, should these handlers only be installed if the kanji ROM is present and large enough (e.g. install them on start after doing the check)?

Reply:
The handlers will now only be installed when a kanji region is present.